src/corecel/data/UserParamsRegistry.hh

```diff
@@ -0,0 +1,97 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file corecel/data/UserParamsRegistry.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <memory>
+
+#include "UserInterface.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Manage user-added parameter classes.
+ *
+ * An instance of this class can be added to shared problem data so that users
+ * can share arbitrary information between parts of the code and create
+ * independent state data for each stream.
+ */
+class UserParamsRegistry
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using SPParams = std::shared_ptr<UserParamsInterface>;
+    using SPConstParams = std::shared_ptr<UserParamsInterface const>;
+    //!@}
+
+  public:
+    // Default constructor
+    UserParamsRegistry() = default;
+
+    //// CONSTRUCTION ////
+
+    //! Get the next available ID
+    UserId next_id() const { return UserId(params_.size()); }
+
+    // Register user parameters
+    void insert(SPParams params);
+
+    //! Get the number of defined params
+    UserId::size_type size() const { return params_.size(); }
+
+    // Access params at the given ID
+    inline SPParams const& at(UserId);
+    inline SPConstParams at(UserId) const;
+
+    // Get the label corresponding to user params
+    inline std::string const& id_to_label(UserId id) const;
+
+    // Find the ID corresponding to an label
+    UserId find(std::string const& label) const;
```

Comment:
I think we already discussed a similar feedback in a previous PR: the `template<T> find` for `unordered_map` is implemented in C++20 but not 17, so this would convert strings to views back to strings to do the search. (For `std::map` surprisingly it was available since 14).